+angstrom.0.7.0

seliopou opened the pull request.

seliopou commented:
@andreas, @darrenldl, @dinosaure, @Leonidas-from-XIV: there are some breaking changes in this release. Specifically, `parse_only` is going away and should be replaced with `parse_string`.

camelus commented:
##### :white_check_mark: All lint checks passed <small>844f5c6f70d3f43b2ef5cf4df7e236097b27e798</small>

* These packages passed lint tests: `angstrom-async.0.7.0`, `angstrom-lwt-unix.0.7.0`, `angstrom-unix.0.7.0`, `angstrom.0.7.0`, `graphql.0.1.0`, `graphql.0.2.0`, `nosetup.0.8.1`, `osbx.1.0.0`, `osbx.1.0.1`, `osbx.1.1.1`, `osbx.1.2.1`, `osbx.1.2.2`, `osbx.1.2.3`, `type-beat.0.1`


---


##### :white_check_mark: Installability check (7458 &rarr; 7462)

* new installable packages (4): `angstrom.0.7.0` `angstrom-async.0.7.0`
  `angstrom-lwt-unix.0.7.0` `angstrom-unix.0.7.0`

seliopou commented:
Build failures are related to to graphql, whose maintainer was already notified of the change as part of this PR.

seliopou commented:
In addition, it looks like some builds are running on `4.02.3` which is not supported by angstrom any more.

andreas commented:
It looks like installing `angstrom-async.0.7.0` fails because `angstrom.0.6.0` is installed. Should `angstrom-async.0.7.0` depend on `angstrom.0.7.0`?

seliopou commented:
@andreas thanks for pointing that out.

I've pushed a patch to change the `opam` file for angstrom-async. I'll change the actual file in the source distribution in the next `0.7.1` release.

seliopou commented on packages/osbx/osbx.1.0.0/opam:
```diff
@@ -11,7 +11,7 @@ depends: [
   "stdint" {build}
   "nocrypto" {build}
-  "angstrom" {build & < "0.7.0"}
+  "angstrom" {build & >= "0.2.0" & < "0.7.0"}
```
@darrenldl It looks like seqbox needs a lower-bound constraint to ensure that the `BE` module is available.

darrenldl replied:
@seliopou Many thanks for keeping everyone updated!

aantron commented:
Hey @seliopou, thanks for the cc. Could you briefly say why the constraint from `lwt.3.1.0` on `ocaml-migrate-parsetree`?

seliopou commented:
It's specifically to get the 4.06 builds working. Without the constraint ,OPAM selects `ocaml-parse-tree.1.0.4`, which has a bug in the `ast_406.ml` that results in a type error. If you scroll down to the bottom of [this log file](https://ci.ocaml.io/log/saved/docker-build-510b380165c08d24798056d45caa2011/4a2f7bbe3912f57149706e6418ec62b12f67ad97), you can see how it fails.

Not sure when 4.06 compatibility became a requirement for new OPAM packages, but I think it may be jumping the gun a bit.

aantron commented:
Thanks. I would think, then, it should be more appropriate to constrain `ocaml-migrate-parsetree` itself with a constraint `avaialble: ocaml-version < "4.06"` for all the broken versions. That will solve the problem for all users of OMP, instead of just Lwt, and permit Lwt to continue triggering installation of older versions of OMP on compiler versions before 4.06, should some constraints make that necessary for some users. cc @let-def

seliopou commented:
Sounds good, and it looks like those changes to ocaml-migrate-parse-tree are already in master... this branch was just a bit behind.

Reverted the change to lwt, let's see what happens.

![2f9](https://user-images.githubusercontent.com/25269/31579806-4c216f18-b10c-11e7-9d23-54322ff208ad.gif)

aantron commented:
:D

Great, thanks for taking a look.

seliopou commented on packages/osbx/osbx.1.0.0/opam:
```diff
@@ -18,3 +18,4 @@ depends: [
   "ctypes-foreign" {build}
   "cmdliner" {build}
 ]
+available: [ ocaml-version >= "4.03.0" ]
```
Had to add this because 4.02.3 builds were failing with a syntax error, which you can see [here](https://travis-ci.org/ocaml/opam-repository/jobs/287979507).

seliopou commented:
@darrenld looks like the `osbx.1.0.1` is breaking the build on OSX, though I'm not sure if that's considered a blocker any more.

darrenldl commented:
@seliopou Yep `osbx.1.0.0` and `osbx.1.0.1` were using FFI along with jbuilder, which was a bit messy.
And yep `osbx.1.0.0` was breaking OSX builds when it was out as well due to FFI stuff(this was fixed in `1.0.1` but FFI was no longer needed after `1.0.1`).

seliopou commented on packages/osbx/osbx.1.0.0/opam:
```diff
@@ -18,4 +18,4 @@ depends: [
   "cmdliner" {build}
 ]
-available: [ ocaml-version >= "4.03.0" ]
+available: [ ocaml-version >= "4.03.0" & os != "darwin" ]
```
@darrenldl Added the darwin constraint.

seliopou commented on packages/sexplib/sexplib.v0.9.0/opam:
```diff
@@ -10,7 +10,7 @@ build: [
 depends: [
   "base"     {= "v0.9.0"}
-  "jbuilder" {build & >= "1.0+beta4"}
+  "jbuilder" {build & >= "1.0+beta4" & < "1.0+beta12"}
```
@trefis @hhugo FYI, this is to handle the removal of `per_file` in beta12.

trefis replied:
I have the vague feeling that we should never have such upper bound constraints on jbuilder if we don't want to make it a nightmare for people.

Let's say you revert that and I do a minor release of sexplib?

gasche replied:
We need a minor release of sexplib for 4.06 compatibility anyway :-) But if the build of sexplib.0.9.0 is broken under future jbuilder releases, isn't it right to have a constraint? Your minor release will be sexplib.0.9.1 anyway, right? (Or are you suggesting to add the constraint back, but only after the minor release is out, so that people have a transition path by updating sexplib first?)

trefis replied:
> We need a minor release of sexplib for 4.06 compatibility anyway :-)

![:|](http://parce-q.eu/l.jpg)

> But if the build of sexplib.0.9.0 is broken under future jbuilder releases, isn't it right to have a constraint? Your minor release will be sexplib.0.9.1 anyway, right? 

That is correct.
I was indeed hoping to release sexplib before this PR gets merged.

seliopou commented:
![af51fd796fcb0b4b76577c345b96e04c_1423875056_cropped](https://user-images.githubusercontent.com/25269/31588367-b214603e-b1be-11e7-9195-166de8e0c18b.png)

seliopou commented:
@avsm @samoht @yomimono et al, cstruct before 3.0.0 is not 4.06.0 compatible because the code generated by Oasis will not compile with safe-string.

I added constraints where appropriate in 37bb7aaf5b223d49667c03fffabb4c1d5458fbbd but this is gonna be a huge bummer since it'll break a lot of packages in the repo.

... plus I have to wait for another one of these mega-builds.

![tenor](https://user-images.githubusercontent.com/25269/31591106-fac8687c-b1e9-11e7-85e6-595d04f6883c.gif)

seliopou commented:
oasis.0.4.5 was the last version that included the unsafe string operation. Fixed by commit ocaml/oasis@7c1de11e, I'm pretty sure.

seliopou commented:
Do these mega-tests get canceled every time I push? Because if it doesn't this is gonna take a while.

![gob](https://user-images.githubusercontent.com/25269/31591528-935c01a8-b1ee-11e7-9e69-0b16bac4f5cb.gif)

gasche commented:
@selioupou: my own testing reports that currently (without your PR), the following versions of cstruct fail to build on 4.06.0:
- anything between 0.4.0 and 1.2.0
- 2.3.2

In particular, your commit https://github.com/ocaml/opam-repository/pull/10286/commits/37bb7aaf5b223d49667c03fffabb4c1d5458fbbd looks like a good idea to me; in particular because all cstruct versions <2.3.3 come with a configure-time test that fails for unsupported OCaml versions, including 4.06 (see https://github.com/mirage/ocaml-cstruct/issues/180 ). On the other hand, I am not sure of why you did the partial revert https://github.com/ocaml/opam-repository/pull/10286/commits:
- as far as I can tell these versions (except for maybe 2.4.1) cannot support 4.06, so the bound is correct
- at least 2.3.2 can currently attempt a build under 4.06 and will fail, so the bound is useful

Do you know if other bounds you changed as part of your PR imply those bounds, is that the reason for the revert?

gasche commented:
I general I think this PR is very nice but there are a bit too much things in it -- I'm not confident it is safe or that it is easy enough to assess its correctness. Would it be possible to separate the part that  concern angstrom directly (the new release, and any change to the previous releases' metadata), and the changes that fix other projects, with the changes with other projects separated into smaller PRs?

avsm commented:
The mega tests don't currently get cancelled. A smaller set of PRs will get through them more easily (e.g. the cstruct changes); I can manually restart the CI if it's super blocked up.  One blocker I'm dealing with at the moment is that the 4.06 images are out of date due to the opam/jbuilder build change that is blocking the rebuilds. Fixing that now...

seliopou commented:
I've broken this PR up into #10515 and #10516. Once those get merged I'll update this PR with a merge from master, at which point hopefully angstrom will be good to go.

seliopou commented:
I've rebased this on top of the current master and cleaned up some reverts in the commit history. I was hoping to get #10516 merged beforehand but don't want to hold this up too much longer.

seliopou commented:
Ok, the last thing that remains not broken out is the version constraint on sexplib before the `v0.9.x` versioning scheme. I'll leave it unless anybody objects.

gasche commented:
That works for me if the CI passes.

seliopou commented:
![plz plz plz](https://media.tenor.com/images/bbc2e3363caf06e4df3e9b4c5f9d73f5/tenor.gif)

gasche commented:
The CI reports failures in four different places.

The Travis CI catches a `type-beat` failure under both 4.04:

```
#=== ERROR while installing type-beat.0.1 =====================================#
[...]
# + ocamlfind ocamlc -c -g -bin-annot -safe-string -warn-error +1..49 -w A-4-41-44 -package angstrom -I lib -o lib/typeBeat_rfc822.cmo lib/typeBeat_rfc822.ml
# File "lib/typeBeat_rfc822.ml", line 153, characters 4-8:
# Error: The type of this expression,
#        _[> `Atom of string | `String of string ] Angstrom.t,
#        contains type variables that cannot be generalized
# Command exited with code 2.
```

ci.ocaml.io runs a bit crazy (because of the many reverse-dependencies of the cstruct changes, I suppose), but it finds -safe-string issues in [base.0.9.3](https://ci.ocaml.io/log/saved/docker-build-864523959e7fbeb6ce270db3c0eee40a/475fae0503d5719749ede6caed0ac0b2d6ef4a45) and [angstrom-lwt-unix](https://ci.ocaml.io/log/saved/docker-build-d6628a8fa67b8e0e4d4355da734d88bd/ead6c03b913becb60607c793bafc97d699b9f219).

gasche commented:
@seliopou, what do you think is best to do? Frankly I think that this works well enough now that we could consider merging it -- and maybe iterate to refine some bounds later. But maybe you want to do things differently? (I wish you'd have split the sexplib change because it would already be merged and this would help for other packages.)

dinosaure commented:
Hmmhmm, about `type-beat`, this lib is ready for `angstrom.0.7` and wait a merging of this PR for a new release. Please, don't consider to block this situation because `type-beat.0.1` does not compile, nobody use it (yet ?).

avsm commented:
This has been a crazy epic effort. Thank you for all the revisions and reviews! I've looked through the logs, and this is good enough to merge now.  The remaining safe-string issues in 4.06 will be far easier to fix in trunk once this is in place.

avsm commented:
You may want to announce this on <https://discuss.ocaml.org>, where we have a Community category and an `announce` tag for this purpose.
